include/zephyr/mgmt/osdp.h

```diff
@@ -202,40 +215,248 @@ struct osdp_cmd {
 	sys_snode_t node;
 	enum osdp_cmd_e id;
 	union {
-		struct osdp_cmd_led    led;
+		struct osdp_cmd_led led;
 		struct osdp_cmd_buzzer buzzer;
-		struct osdp_cmd_text   text;
+		struct osdp_cmd_text text;
 		struct osdp_cmd_output output;
 		struct osdp_cmd_comset comset;
 		struct osdp_cmd_keyset keyset;
+		struct osdp_cmd_mfg mfg;
+	};
+};
+
+/* ------------------------------- */
+/*          OSDP Events            */
+/* ------------------------------- */
+
+/**
+ * @brief Various card formats that a PD can support. This is sent to CP
+ * when a PD must report a card read.
+ */
+enum osdp_event_cardread_format_e {
+	OSDP_CARD_FMT_RAW_UNSPECIFIED,
+	OSDP_CARD_FMT_RAW_WIEGAND,
+	OSDP_CARD_FMT_ASCII,
+	OSDP_CARD_FMT_SENTINEL
+};
+
+/**
+ * @brief OSDP event cardread
+ *
+ * @param reader_no In context of readers attached to current PD, this number
+ *        indicated this number. This is not supported by LibOSDP.
+ * @param format Format of the card being read.
+ *        see `enum osdp_event_cardread_format_e`
+ * @param direction Card read direction of PD 0 - Forward; 1 - Backward
+ * @param length Length of card data in bytes or bits depending on `format`
+ *        (see note).
+ * @param data Card data of `length` bytes or bits bits depending on `format`
+ *        (see note).
+ *
+ * @note When `format` is set to OSDP_CARD_FMT_RAW_UNSPECIFIED or
+ * OSDP_CARD_FMT_RAW_WIEGAND, the length is expressed in bits. OTOH, when it is
+ * set to OSDP_CARD_FMT_ASCII, the length is in bytes. The number of bytes to
+ * read from the `data` field must be interpreted accordingly.
+ */
+struct osdp_event_cardread {
+	int reader_no;
+	enum osdp_event_cardread_format_e format;
+	int direction;
+	int length;
+	uint8_t data[OSDP_EVENT_MAX_DATALEN];
+};
+
+/**
+ * @brief OSDP Event Keypad
+ *
+ * @param reader_no In context of readers attached to current PD, this number
+ *                  indicated this number. This is not supported by LibOSDP.
+ * @param length Length of keypress data in bytes
+ * @param data keypress data of `length` bytes
+ */
+struct osdp_event_keypress {
+	int reader_no;
+	int length;
+	uint8_t data[OSDP_EVENT_MAX_DATALEN];
+};
+
+/**
+ * @brief OSDP Event Manufacturer Specific Command
+ *
+ * Note: OSDP spec v2.2 makes this structure fixed at 4 bytes. LibOSDP allows
+ * for some additional data to be passed in this command using the data and
+ * length fields. To be fully compliant with the specification, set the length
+ * field to 0.
+ *
+ * @param vendor_code 3-bytes IEEE assigned OUI of manufacturer
+ * @param command 1-byte reply code
+ * @param length Length of manufacturer data in bytes (optional)
+ * @param data manufacturer data of `length` bytes (optional)
+ */
+struct osdp_event_mfgrep {
```

Comment:
Relevant events/cmds should have the same size of members. 
This comes out from the fact that event can be used as response on POLL cmd. 
In this situation event (provided in ephemeral data) is directly casted to relevant command struct so memory layout of cmd/event structs should be the same.

Reply:
Thanks for the review. In PD, `CMD_MFG` copies the response from the app as `struct osdp_cmd` into ephemeral_data ([here](https://github.com/goToMain/zephyr/blob/osdp/subsys/mgmt/osdp/src/osdp_pd.c#L499)) and in `REPLY_MFGREP`, we cast it back to `struct osdp_cmd`.

So there is no problem with that code as it is now but still, I agree it can be a problem in future if we want to allow PDs to report unsolicited MFGREPs. I'll make `osdp_event_mfgrep` look like `osdp_cmd_mfg`